### PR TITLE
OTC-111: Changing user roles

### PR DESCRIPTION
**Pull request by bartekkaminski**
TASK:
https://openimis.atlassian.net/browse/OTC-111

DATABASE:
https://github.com/openimis/database_ms_sqlserver/pull/32

### CODE REVIEW

File: IMIS/User.aspx.vb [hunk 1]
```diff
@@ -204,21 +217,24 @@ Partial Public Class User
         Next
         Return checked
     End Function
-    Private Sub B_SAVE_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles B_SAVE.Click
-        If txtPassword.Text <> String.Empty Then
```
Comment (dragos-dobre) [inline IMIS/User.aspx.vb:208]: Why did you remove this validation? If no password is given, then the password should not be updated. Is it still the case?
Also, the password straight validation from configuration is not executed anymore, right?

File: IMIS/User.aspx.vb [hunk 2]
```diff
@@ -290,4 +333,30 @@ Partial Public Class User
         Response.Redirect("FindUser.aspx?u=" & txtLoginName.Text)
     End Sub
 
+    Private Function IsValidPassword() As Integer
```
Comment (dragos-dobre) [inline IMIS/User.aspx.vb:336]: Check the password validation from configuration that was developed by you or Arek. This should not be here.

File: IMIS_BL/UsersBL.vb [hunk 3]
```diff
@@ -949,4 +952,32 @@ Public Class UsersBL
...
+        Dim ds As New DataSet
+        ds = User.GetUserDistricts(CurrenctUserID, SelectedUserID)
+        Dim dtSelectedUserDistricts As DataTable = ds.Tables("SelectedUserDistricts")
+        Dim dtCurrentUserDistricts As DataTable = ds.Tables("CurrentUserDistricts")
+
+        Dim dtSelectedUserRegions As DataTable = ds.Tables("SelectedUserRegions")
+        Dim dtCurrentUserRegions As DataTable = ds.Tables("CurrentUserRegions")
+
+        Dim Users As New IMIS_DAL.UsersDAL
+        If dtCurrentUserRegions.Rows.Count = 1 Then
```
Comment (dragos-dobre) [inline IMIS_BL/UsersBL.vb:966]: Please combine the IFs.

File: IMIS_DAL/RoleDAL.vb [hunk 4]
```diff
@@ -34,7 +34,7 @@ Public Class RoleDAL
         Dim sSQL As String = String.Empty
         Dim data As New ExactSQL
 
-        sSQL = "SELECT [RoleID],[RoleUUID],[RoleName],CASE WHEN [IsSystem] > 0 THEN 'True' Else 'False' END System,isSystem,ISNULL(AltLanguage,RoleName) AltLanguage,"
+        sSQL = "SELECT [RoleID],[RoleName],CASE WHEN [IsSystem] > 0 THEN 'True' Else 'False' END System,isSystem,ISNULL(AltLanguage,RoleName) AltLanguage,"
```
Comment (dragos-dobre) [inline IMIS_DAL/RoleDAL.vb:37]: You removed the RoleUUID. I'm sure this is important as you need to navigate with UUID. Maybe this is why the bug OTC-121?

File: IMIS_DAL/RoleDAL.vb [hunk 5]
```diff
@@ -185,8 +192,8 @@ Public Class RoleDAL
     Public Function GetSystemRoles(RoleID) As DataTable
         Dim sSQL As String = String.Empty
         Dim data As New ExactSQL
-        sSQL = "SELECT RoleId,RoleName FROM tblRole"
-        sSQL += " WHERE (IsSystem & @isSystem) > 0 AND ValidityTo IS NULL"
+        sSQL = "SELECT 0,0,RoleId,getdate(),Getdate(),0,0,1 FROM tblRole"
```
Comment (dragos-dobre) [inline IMIS_DAL/RoleDAL.vb:195]: I don't understand this line. Doesn't make any sense with the tblRole structure. Was it like this in Exact version?

**Comment by dragos-dobre** [comment]
@kmadej can you see with Damian or Dawid to finalize this PR? thanks.

**Comment by kmadej** [comment]
ok, @druchniewicz please look into this

**Comment by delcroip** [comment]
@dborowiecki please check those comments, the webapp cannot work without

File: IMIS_DAL/RoleDAL.vb (hunk 5 above)
Comment (dborowiecki) [reply IMIS_DAL/RoleDAL.vb:195]: It was like this in Exact version. This query mocks record from tblUserRole. If query is in other shape then following error is raised: 

> Trying to pass a table-valued parameter with 2 column(s) where the corresponding user-defined table type requires 8 column(s).

File: IMIS_DAL/RoleDAL.vb (hunk 4 above)
Comment (dborowiecki) [reply IMIS_DAL/RoleDAL.vb]: I've added it again. After restoring RoleUUID, list of Roles in FindProfile.aspx page is visible again, so it most likely caused the OTC-121 bug.

File: IMIS/User.aspx.vb [hunk 6]
```diff
@@ -204,21 +217,24 @@ Partial Public Class User
...
-                Exit Sub
-            End If
-            If txtPassword.Text <> txtConfirmPassword.Text Then
-                lblMsg.Text = imisgen.getMessage("V_CONFIRMPASSWORD")
-                Exit Sub
-            End If
-            eUsers.DummyPwd = txtPassword.Text
-        End If
 
+    Public Function GetJson(ByVal dt As DataTable) As String
```
[the remark this answers is not in the document]
Comment (dborowiecki) [reply IMIS/User.aspx.vb]: It wasn't referenced anywhere. Removed.

File: IMIS/User.aspx.vb (hunk 2 above)
Comment (delcroip) [reply IMIS/User.aspx.vb:343]: @dborowiecki could you check why password check in here ?  there is the same function in IMIS/General.vb

File: IMIS_BL/UsersBL.vb (hunk 3 above)
Comment (delcroip) [reply IMIS_BL/UsersBL.vb:966]: @dborowiecki please use logic AND in the IF instead of nested ifs

File: IMIS_DAL/RoleDAL.vb (hunk 4 above)
Comment (delcroip) [reply IMIS_DAL/RoleDAL.vb]: @dborowiecki I don't see it on the file, are you sure you commited it ?

File: IMIS/User.aspx.vb (hunk 2 above)
Comment (dborowiecki) [reply IMIS/User.aspx.vb]: @delcroip I've used function from IMIS/General.vb and removed additional validation.

File: IMIS/Role.aspx.vb [hunk 7]
```diff
@@ -38,12 +38,17 @@
...
         tvRoleRights4.ExpandAll()
-        If eRole.IsSystem Then
+        If eRole.IsSystem Or eRole.ValidityTo IsNot Nothing Then
 
             txtRoles.Enabled = False
             txtAltLanguage.Enabled = False
             chkIsSystem.Enabled = False
-            chkIsBlocked.Enabled = True
+            If eRole.ValidityTo IsNot Nothing Then
+                chkIsBlocked.Enabled = False
```
[the remark this answers is not in the document]
Comment (dborowiecki) [reply IMIS/Role.aspx.vb:47]: @delcroip You're right about the block button. I've thought that this condition: 
```
If eRole.ValidityTo IsNot Nothing Then
                chkIsBlocked.Enabled = False
```
didn't have any effect since outdated roles are not displayed. I've missed that they can be shown by checking 'Historical' option. In that case disabling 'Blocked' checkbox in obsolete roles make sense (they cannot be saved anyway).

There was also an issue with blocking system roles (IsSystem was set to 1 and therefore EnrollmentOfficer was always
assigned after updating system role), I've fixed it in [last commit](https://github.com/openimis/web_app_vb/pull/40/commits/868751e8599d231e6012a20d350aa97f92074a7c).

File: IMIS_DAL/RoleDAL.vb (hunk 5 above)
Comment (dborowiecki) [reply IMIS_DAL/RoleDAL.vb:199]: @delcroip I've looked into this and previously xAttribute was used, but due to adding new column 'Assign' it was 
changed to xtblUserRole type.

File: IMIS/User.aspx.vb (hunk 6 above)
[the remark this answers is not in the document]
Comment (dborowiecki) [reply IMIS/User.aspx.vb]: @delcroip I think occurrences you've mention are JavaScript ones, and above GetJson comes from VB class, 
so although the names are identical, the above function is not used.

**Review by delcroip — APPROVED** [review]

**Review by delcroip — APPROVED** [review]
